File: LeetCode/120. Triangle.js
// https://leetcode.com/problems/triangle/description/

// Перебираем ряды, начиная с предпоследнего и до самого первого
// Для каждого элемента ряда сохраняем минимальное значение: элемент + минимальный из двух элементов под ним
const minimumTotal = (triangle) => {
  const numberOfRows = triangle.length;

  const temp = [...triangle[numberOfRows - 1]];

  for (let i = numberOfRows - 2; i >= 0; i--) {
    const row = triangle[i];

    for (let k = 0; k < row.length; k++) {
      temp[k] = row[k] + Math.min(temp[k], temp[k + 1]);
    }
  }

  return temp[0];
};

console.log(minimumTotal([[2], [3, 4], [6, 5, 7], [4, 1, 8, 3]])); // 11
console.log(minimumTotal([[-10]])); // -10

// Рекурсия, не проходит тесты по времени
// const minimumTotal = (triangle) => {
//   let min = +Infinity;

//   const inner = (index, level, sum) => {
//     if (level >= triangle.length) {
//       if (sum < min) {
//         min = sum;
//       }

//       return;
//     }

//     const current = triangle[level][index];

//     inner(index, level + 1, sum + current);
//     inner(index + 1, level + 1, sum + current);
//   };

//   inner(0, 0, 0);

//   return min;
// };

File: LeetCode/120. Triangle.py
# https://leetcode.com/problems/triangle/description/


# Перебираем ряды, начиная с предпоследнего и до самого первого
# Для каждого элемента ряда сохраняем минимальное значение: элемент + минимальный из двух элементов под ним
class Solution:
    def minimumTotal(self, triangle) -> int:
        if len(triangle) == 1:
            return triangle[0][0]

        numberOfRows = len(triangle)

        temp = list(triangle[numberOfRows - 1])

        for row in triangle[numberOfRows - 2 :: -1]:

            for index, value in enumerate(row):
                temp[index] = value + min(temp[index], temp[index + 1])

        return temp[0]
